### PR TITLE
Yufei - Add a mouseover text edit button near 'Total Time'

### PR DESCRIPTION
**Pull request by yufeiStella**
# Description
Screen Shot 2023-07-21 at 16 31 39

## Related PRS (if any):
To test this backend PR you need to checkout the [#1030](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1030) frontend PR.

## Main changes explained:
- Create new modal mouseoverText.js
- Create new controller mouseoverTextController.js
- Create new routes mouseoverTextRoute.js
- Update startup/routes to add mouseoverText api

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. log in as an owner user and verify that the mouseover text appears when you move your mouse over both 'Total Time' and the numbers on the Leaderboard. Some numbers are blue (indicating intangible hours logged), and some numbers are black. Mouseover text will appear on both types of numbers.
4. Only the owner can see the edit button next to 'Total Time'. Click on it and enter the new text to update the mouseover text. Please note that it may take some time to respond, as shown in my screen recording.
5. Wait a few seconds for a green pop-up box to appear, showing the message "Mouseover Text updated!" Then you will see the newly updated mouseover text when you move your mouse over both 'Total Time' and the numbers. If you don't immediately see the changes, just wait a few more seconds, as it may take some time for the updates to respond.
6. Hard refresh the page to see the change is globally. The same thing is if you don't immediately see the changes, just wait a few more seconds, as it may take some time for the updates to respond.
7. log in as any other role to verify the mouseoverText updates globally and they do not see the edit button.

## Screenshots or videos of changes:
Log in as an owner:
Screen Shot 2023-07-21 at 19 43 12

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/4ceff3aa-ae86-4ba2-8dfa-096808022cb5

Log in as any other user except owner:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/46e3df42-b839-4469-bd34-afc8113ecf54

## Note:
Sometimes, you may need to wait for the page to respond and display the mouseover text. I encountered a similar situation as shown in my screenshot recording. When you move your mouse to another place where there should be mouseover text, you may experience the same situation. Just be patient and wait.

### CODE REVIEW

**Comment by n0vay** [comment]
How do I create a owner account?

**Review by AuroraLovesWhales — APPROVED** [review]
Hi @yufeiStella. I have left my comment in the related frontend PR. The functionality is working as intended. Great job!

**Review by shiva2096 — APPROVED** [review]
Hi @yufeiStella 

Reviewed this PR with frontend PR 1030, working as expected .

**Comment by AnishPandita** [comment]
Hi @yufeiStella I can confirm it works as intended.

Screenshot 2023-07-24 at 5 25 36 PM
Screenshot 2023-07-24 at 5 31 50 PM
Screenshot 2023-07-24 at 5 33 07 PM

**Review by DavidC0126 — APPROVED** [review]
Hi @yufeiStella,

Reviewed this PR with frontend [PR#1030](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1030#pullrequestreview-1544612531), working as expected.

**Review by sxiong5 — APPROVED** [review]
Hey, I left a comment in [HighestGoodNetworkApp#1030](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1030). The PR worked as expected.

**Review by veronicacheng2 — APPROVED** [review]
Hi @yufeiStella, I have tested together with the Frontend, and they are working as intended


When logged in as Owner, there will be a text edit button:
Logged in as Owner























Clicking the edit text button:
Edit mouseover text


Hovering to total time:
Hover onto total time



No edit text button when logged in as volunteer:
Logged in as Volunteer

**Comment by Shreyj14** [comment]
Hi @yufeiStella

Reviewed this PR with frontend PR 1030, working as expected .

**Review by Shreyj14 — APPROVED** [review]

**Comment by iTvX** [comment]
Tested with the front end, I have review on the front end.

`I have checked your PR. The owner can change the text, and it looks good. But when I switch accounts (not the owner's account), I cannot see the text. It seems like it's not global.

`

**Comment by mounica-dingari-by** [comment]
Hi @yufeiStella , I tried to test this feature, multiple times. But, I couldn't as the app hangs a lot once I switched to yufei_mouseover_text_controller branch. I'm not sure, if it's because this branch or not. I'm including the screenshots below.

Screenshot from 2023-07-27 21-37-07
Screenshot from 2023-07-27 21-37-13

File: src/controllers/mouseoverTextController.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    const updateMouseoverText = function (req, res) {
+        // if (req.body.requestor.role !== 'Owner') {
+        //     res.status(403).send('You are not authorized to update mouseoverText!');
+        // }
+        const { id } = req.params;
+
+        return MouseoverText.findById(id, (error, mouseoverText) => {
+            if (error || mouseoverText === null) {
+                res.status(400).send('No mouseoverText found');
```
Comment (iTvX) [inline src/controllers/mouseoverTextController.js:26]: 400 means bad request, but in this case the error could be database error. I suggest use error 500.

Comment (yurisokolovicz) [inline src/controllers/mouseoverTextController.js:26]: For improving error handling, I think the error 404 code could be used, along with a more intuitive error message that could aid in debugging.

Example:    

res.status(404).send('MouseoverText not found with the given ID' );

**Review by yurisokolovicz — CHANGES_REQUESTED** [review]
Dear @yufeiStella,

The code looks good overall. I think making minor corrections in error handling could help in further debugging. Excellent work!

**Comment by yufeiStella** [comment]
> Tested with the front end, I have review on the front end.
> 
> `I have checked your PR. The owner can change the text, and it looks good. But when I switch accounts (not the owner's account), I cannot see the text. It seems like it's not global. `
> 
>  256637192-cd5d1198-0e63-4f84-a35d-92400f833841.mp4

Hi @iTvX! It's all about slow loading. For me, usually I just reload it again since it's too slow.  You can also find that we are talking the problem in slack(code channel). 
Screen Shot 2023-08-02 at 00 28 56

File: src/controllers/mouseoverTextController.js (hunk 1 above)
Comment (yufeiStella) [reply src/controllers/mouseoverTextController.js]: Updated! @yurisokolovicz, could you please re-review the PR and approve it if you agree with it? Thank you

Comment (yufeiStella) [reply src/controllers/mouseoverTextController.js]: Thank you! I use error 500 now.

**Review by yurisokolovicz — APPROVED** [review]
Perfect!

**Comment by beblicarl** [comment]
Tested, everything works as expected and the code is great!

**Review by beblicarl — APPROVED** [review]

**Pull request merged**